# Machine Learning - Continuous Assessment - Class Group: TU060 <a class="tocSkip">

# MSc in Computer Science - Data Science (Part Time) <a class="tocSkip">

# Student Name: Ciaran Finnegan  <a class="tocSkip">

# Student No: D21124026 <a class="tocSkip">

#### Student EMail: D21124026@mytudublin.ie <a class="tocSkip">

## May 2022<a class="tocSkip">

# 

# H&M Personalised Fashion Recommendations <a class="tocSkip">


#### This project takes the datasets provided by Kaggle for the H&M Personalised Fashion Recommendations competition.  <a class="tocSkip">

Ipso forum...

## <a class="tocSkip">

#### The CA follows a n-part structure. <a class="tocSkip">

Section 1: Ipso forum...

Section 2: Ipso forum...

Section 3: Ipso forum...


### <a class="tocSkip">

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Jupyter-Notebook-Set-Up" data-toc-modified-id="Jupyter-Notebook-Set-Up-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Jupyter Notebook Set Up</a></span><ul class="toc-item"><li><span><a href="#Import-Libraries" data-toc-modified-id="Import-Libraries-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Import Libraries</a></span><ul class="toc-item"><li><span><a href="#General-Utility-Libraries" data-toc-modified-id="General-Utility-Libraries-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>General Utility Libraries</a></span></li><li><span><a href="#Panda-+-Numpy" data-toc-modified-id="Panda-+-Numpy-1.1.2"><span class="toc-item-num">1.1.2&nbsp;&nbsp;</span>Panda + Numpy</a></span></li><li><span><a href="#Matplotlib-Library" data-toc-modified-id="Matplotlib-Library-1.1.3"><span class="toc-item-num">1.1.3&nbsp;&nbsp;</span>Matplotlib Library</a></span></li><li><span><a href="#Plotly-Libraries" data-toc-modified-id="Plotly-Libraries-1.1.4"><span class="toc-item-num">1.1.4&nbsp;&nbsp;</span>Plotly Libraries</a></span></li></ul></li></ul></li><li><span><a href="#Project-Data-Import,-Analysis-and-Preparation" data-toc-modified-id="Project-Data-Import,-Analysis-and-Preparation-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Project Data Import, Analysis and Preparation</a></span><ul class="toc-item"><li><span><a href="#Load-and-Sample-Kaggle-H&amp;M-Datasets" data-toc-modified-id="Load-and-Sample-Kaggle-H&amp;M-Datasets-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Load and Sample Kaggle H&amp;M Datasets</a></span><ul class="toc-item"><li><span><a href="#Set-Up-File-Paths" data-toc-modified-id="Set-Up-File-Paths-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Set Up File Paths</a></span></li><li><span><a href="#H&amp;M-Transaction-Dataset" data-toc-modified-id="H&amp;M-Transaction-Dataset-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>H&amp;M Transaction Dataset</a></span></li><li><span><a href="#H&amp;M-Articles-Dataset" data-toc-modified-id="H&amp;M-Articles-Dataset-2.1.3"><span class="toc-item-num">2.1.3&nbsp;&nbsp;</span>H&amp;M Articles Dataset</a></span></li><li><span><a href="#H&amp;M-Customers-Dataset" data-toc-modified-id="H&amp;M-Customers-Dataset-2.1.4"><span class="toc-item-num">2.1.4&nbsp;&nbsp;</span>H&amp;M Customers Dataset</a></span></li></ul></li><li><span><a href="#Exploratory-Data-Analysis" data-toc-modified-id="Exploratory-Data-Analysis-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Exploratory Data Analysis</a></span><ul class="toc-item"><li><span><a href="#H&amp;M-Sampled-Transaction-Dataset" data-toc-modified-id="H&amp;M-Sampled-Transaction-Dataset-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>H&amp;M Sampled Transaction Dataset</a></span></li><li><span><a href="#H&amp;M-Sampled-Customer-Dataset" data-toc-modified-id="H&amp;M-Sampled-Customer-Dataset-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>H&amp;M Sampled Customer Dataset</a></span></li><li><span><a href="#H&amp;M-Sampled-Articles-Dataset" data-toc-modified-id="H&amp;M-Sampled-Articles-Dataset-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>H&amp;M Sampled Articles Dataset</a></span></li></ul></li><li><span><a href="#Data-Preparation-and-Feature-Enrichment" data-toc-modified-id="Data-Preparation-and-Feature-Enrichment-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Data Preparation and Feature Enrichment</a></span><ul class="toc-item"><li><span><a href="#Transaction-Data-Feature-Enrichment-(Sample)" data-toc-modified-id="Transaction-Data-Feature-Enrichment-(Sample)-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Transaction Data Feature Enrichment (Sample)</a></span></li><li><span><a href="#Extend-Dataframe-with-Customer-Data" data-toc-modified-id="Extend-Dataframe-with-Customer-Data-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Extend Dataframe with Customer Data</a></span></li><li><span><a href="#Extend-Dataframe-with-Articles-Data" data-toc-modified-id="Extend-Dataframe-with-Articles-Data-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>Extend Dataframe with Articles Data</a></span></li><li><span><a href="#Final-Data-Preparation-Steps" data-toc-modified-id="Final-Data-Preparation-Steps-2.3.4"><span class="toc-item-num">2.3.4&nbsp;&nbsp;</span>Final Data Preparation Steps</a></span></li></ul></li></ul></li><li><span><a href="#Split-Data:-Train-and-Test-Sets" data-toc-modified-id="Split-Data:-Train-and-Test-Sets-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Split Data: Train and Test Sets</a></span></li><li><span><a href="#Model-Preparation" data-toc-modified-id="Model-Preparation-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Model Preparation</a></span></li><li><span><a href="#K-Nearest-Neighbor-Recommendations" data-toc-modified-id="K-Nearest-Neighbor-Recommendations-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>K Nearest Neighbor Recommendations</a></span></li><li><span><a href="#Submission" data-toc-modified-id="Submission-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Submission</a></span></li><li><span><a href="#References" data-toc-modified-id="References-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>References</a></span></li></ul></div>

## Jupyter Notebook Set Up

### Import Libraries

Note: It is assumed that the matplotlib and plotly library has been installed on the user's machine before it is available for import.

```bash
> pip install matplotlib
> pip install plotly
```


#### General Utility Libraries

In [201]:
import os

#### Panda + Numpy

In [202]:
# Python libraries for data manipulation
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import NearestNeighbors

Set up additional libraries for graphical presentations. A number of earlier graphs use Matplotlib. In later sections, Plotly graphs are used.

#### Matplotlib Library

In [203]:
#import matplotlib.pyplot as plt

#### Plotly Libraries

In [204]:
# For interactive graphics and global displays
#import plotly.express as px
#import plotly.graph_objects as go

### <a class="tocSkip">

In [205]:
import warnings
warnings.filterwarnings("ignore")

### <a class="tocSkip">

## Project Data Import, Analysis and Preparation

### Load and Sample Kaggle H&M Datasets

### The assignment starts with a Kaggle H&M Personalised Fashion Recommendations datasets. <a class="tocSkip">

#### Set Up File Paths 

In [206]:
# Set up a Global path file variable to use when loading each dataset
Path_To_Inputs = '../input'

#### H&M Transaction Dataset 

In [207]:
# Load H&M Transaction dataset
df_trxns = pd.read_csv(os.path.join(Path_To_Inputs,'transactions_train.csv'))

In [208]:
# Display Number of Rows, Number of Columns in Transaction dataset prior to Sampling
df_trxns.shape

(31788324, 5)

In [209]:
# Extract random sample of Transactions (without replacement) - 0.1% of records
df_trxns_sample = df_trxns.sample(frac=0.001, replace=False)
## NEW CODE - Check this out
## df_trxns_sample.set_index(inplace = True)

In [210]:
# Display Number of Rows, Number of Columns in Transaction dataset after Sampling
df_trxns_sample.shape

(31788, 5)

In [211]:
# Check out the index
df_trxns_sample.head()

t_dat                                        customer_id  \
4413783   2019-01-01  24df7e5a17b32fb47e413c0909ea38794ebb3b5f48987f...   
15953688  2019-08-27  2a009e5b897a07be7a7c22924c41dc35ae2706363163fc...   
18004153  2019-10-17  42d005d022b0afee473b01ab3ab3fc6ccfea377b520e55...   
15948120  2019-08-26  fb8fc6d3f2f59376d4e7d79273e3aa3278fa5472edc5bf...   
18584603  2019-11-03  4c598e57d6c33608afff9fd0f76a94b4964fc907bd69ee...   

          article_id     price  sales_channel_id  
4413783    706016002  0.033881                 2  
15953688   721932002  0.045746                 2  
18004153   688558002  0.016932                 1  
15948120   331093006  0.025407                 2  
18584603   746775008  0.020322                 2

#### H&M Articles Dataset 

In [212]:
# Read file, located in 'input' folder above this notebook, for articles data - containing meta data on purchasable items.
df_Articles = pd.read_csv(os.path.join(Path_To_Inputs,'articles.csv'))

In [213]:
# Display Number of Rows, Number of Columns in Articles dataset
df_Articles.shape

(105542, 25)

In [214]:
# Display basic file data for the initial Kaggle H&M 'articles' datatset
# df_Articles.head()

In [215]:
# Extract only the H&M articles contained in the sample Transaction dataset 
articleIDs_in_Trxn_Sample  = set(df_trxns_sample["article_id"])

In [216]:
#Count Article Ids in Transaction Sample
#len(articleIDs_in_Trxn_Sample)

In [217]:
# Create dataset that only contains H&M articles in the sample Transaction dataset
df_articles_sample = df_Articles[df_Articles["article_id"].isin(articleIDs_in_Trxn_Sample)]

In [218]:
# Display Number of Rows, Number of Columns in Sample Articles dataset
df_articles_sample.shape

(18361, 25)

#### H&M Customers Dataset 

In [219]:
# Load H&M Customer dataset
df_customers = pd.read_csv(os.path.join(Path_To_Inputs,'customers.csv'))

In [220]:
# Display Number of Rows, Number of Columns in Customers dataset prior to Sampling
df_customers.shape

(1371980, 7)

In [221]:
# Extract only the H&M Customers contained in the sample Transaction dataset 
custIDs_in_Trxn_Sample  = set(df_trxns_sample["customer_id"])

In [222]:
#Count Customer Ids in Transaction Sample
len(custIDs_in_Trxn_Sample)

30485

In [223]:
# Create dataset that only contains H&M Customers from the sample Transaction dataset
df_customers_sample = df_customers[df_customers["customer_id"].isin(custIDs_in_Trxn_Sample)]

In [224]:
# Display Number of Rows, Number of Columns in Sample Customers dataset
df_customers_sample.shape

(30485, 7)

###### 

### Exploratory Data Analysis

### Display key data attributes and distributions in Sampled H&M datasets. <a class="tocSkip">

#### H&M Sampled Transaction Dataset 

In [225]:
# Display basic column information for the initial Kaggle 'transaction_train' datatset - after Sampling of original dataset
df_trxns_sample.head()

t_dat                                        customer_id  \
4413783   2019-01-01  24df7e5a17b32fb47e413c0909ea38794ebb3b5f48987f...   
15953688  2019-08-27  2a009e5b897a07be7a7c22924c41dc35ae2706363163fc...   
18004153  2019-10-17  42d005d022b0afee473b01ab3ab3fc6ccfea377b520e55...   
15948120  2019-08-26  fb8fc6d3f2f59376d4e7d79273e3aa3278fa5472edc5bf...   
18584603  2019-11-03  4c598e57d6c33608afff9fd0f76a94b4964fc907bd69ee...   

          article_id     price  sales_channel_id  
4413783    706016002  0.033881                 2  
15953688   721932002  0.045746                 2  
18004153   688558002  0.016932                 1  
15948120   331093006  0.025407                 2  
18584603   746775008  0.020322                 2

In [226]:
# Display datatype of colums in original H&M Transaction dataset
df_trxns_sample.dtypes

t_dat                object
customer_id          object
article_id            int64
price               float64
sales_channel_id      int64
dtype: object

#### H&M Sampled Customer Dataset 

In [227]:
# Display basic column information for the initial Kaggle 'customers' datatset - after Sampling of original dataset
df_customers_sample.head()

customer_id   FN  Active  \
38   0001d44dbe7f6c4b35200abdb052c77a87596fe1bdcc37...  1.0     1.0   
179  00077dbd5c4a4991e092e63893ccf29294a9d5c46e8501...  NaN     NaN   
195  00080403a669b3b89d1bef1ec73ea466d95e39698d6dde...  NaN     NaN   
289  000d85c6d74511666a082d1fd399760b155a0647911239...  NaN     NaN   
311  000ed4cabda96069b547ae531c66de5f1bf6a7e5b83aa8...  NaN     NaN   

    club_member_status fashion_news_frequency   age  \
38              ACTIVE              Regularly  44.0   
179             ACTIVE                   NONE  23.0   
195             ACTIVE                   NONE  34.0   
289             ACTIVE                   NONE  55.0   
311             ACTIVE                   NONE  27.0   

                                           postal_code  
38   930b19ae7db8abb5a27f4da10217755a7305b4c452f5e0...  
179  3119ea10ffe5ac3419b9127589a61b33e1ae38ecbb997b...  
195  5af3c7af7cfd5b817222d7f04f7f9bf6dbdaa51f3b6ca1...  
289  139717d5fe01b56cd0111e397ecb6f387a92de52bfe354...  
311  e3c5d89fcf647429be1690898acf620736ffb589811d24...

In [228]:
df_customers_sample.dtypes

customer_id                object
FN                        float64
Active                    float64
club_member_status         object
fashion_news_frequency     object
age                       float64
postal_code                object
dtype: object

#### H&M Sampled Articles Dataset 

In [229]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code                  prod_name  product_type_no  \
0   108775015        108775                  Strap top              253   
1   108775044        108775                  Strap top              253   
6   111565001        111565        20 den 1p Stockings              304   
8   111586001        111586  Shape Up 30 den 1p Tights              273   
9   111593001        111593   Support 40 den 1p Tights              304   

  product_type_name  product_group_name  graphical_appearance_no  \
0          Vest top  Garment Upper body                  1010016   
1          Vest top  Garment Upper body                  1010016   
6  Underwear Tights      Socks & Tights                  1010016   
8   Leggings/Tights  Garment Lower body                  1010016   
9  Underwear Tights      Socks & Tights                  1010016   

  graphical_appearance_name  colour_group_code colour_group_name  ...  \
0                     Solid                  9             Black  ...   
1                     Solid                 10             White  ...   
6                     Solid                  9             Black  ...   
8                     Solid                  9             Black  ...   
9                     Solid                  9             Black  ...   

   department_name index_code        index_name index_group_no  \
0     Jersey Basic          A        Ladieswear              1   
1     Jersey Basic          A        Ladieswear              1   
6     Tights basic          B  Lingeries/Tights              1   
8     Tights basic          B  Lingeries/Tights              1   
9     Tights basic          B  Lingeries/Tights              1   

   index_group_name section_no                    section_name  \
0        Ladieswear         16          Womens Everyday Basics   
1        Ladieswear         16          Womens Everyday Basics   
6        Ladieswear         62  Womens Nightwear, Socks & Tigh   
8        Ladieswear         62  Womens Nightwear, Socks & Tigh   
9        Ladieswear         62  Womens Nightwear, Socks & Tigh   

  garment_group_no  garment_group_name  \
0             1002        Jersey Basic   
1             1002        Jersey Basic   
6             1021    Socks and Tights   
8             1021    Socks and Tights   
9             1021    Socks and Tights   

                                         detail_desc  
0            Jersey top with narrow shoulder straps.  
1            Jersey top with narrow shoulder straps.  
6  Semi shiny nylon stockings with a wide, reinfo...  
8  Tights with built-in support to lift the botto...  
9  Semi shiny tights that shape the tummy, thighs...  

[5 rows x 25 columns]

###### 

### Data Preparation and Feature Enrichment

### Construct a dataset upon which to build ML models. <a class="tocSkip">

### Begin with Transaction Dataset (sampled) and merge in customer and article data. <a class="tocSkip">

#### Transaction Data Feature Enrichment (Sample)

In [230]:
# Create new dataset for modelling. Initial basis for new dataset is the sampled transaction dataset.
df_HM_CustTrxn = df_trxns_sample

In [231]:
df_HM_CustTrxn.head()

t_dat                                        customer_id  \
4413783   2019-01-01  24df7e5a17b32fb47e413c0909ea38794ebb3b5f48987f...   
15953688  2019-08-27  2a009e5b897a07be7a7c22924c41dc35ae2706363163fc...   
18004153  2019-10-17  42d005d022b0afee473b01ab3ab3fc6ccfea377b520e55...   
15948120  2019-08-26  fb8fc6d3f2f59376d4e7d79273e3aa3278fa5472edc5bf...   
18584603  2019-11-03  4c598e57d6c33608afff9fd0f76a94b4964fc907bd69ee...   

          article_id     price  sales_channel_id  
4413783    706016002  0.033881                 2  
15953688   721932002  0.045746                 2  
18004153   688558002  0.016932                 1  
15948120   331093006  0.025407                 2  
18584603   746775008  0.020322                 2

In [232]:
# Decompose Datetime information 
df_HM_CustTrxn['t_dat'] = pd.to_datetime(df_HM_CustTrxn['t_dat'], format="%Y-%m-%d")

# Add Column for Transaction Year and convert to integer value
df_HM_CustTrxn['txn_year'] = df_HM_CustTrxn['t_dat'].dt.strftime('%Y')
df_HM_CustTrxn['txn_year'] = df_HM_CustTrxn['txn_year'].astype(str).astype(int)

# Add Column for Transaction Month and convert to integer value
df_HM_CustTrxn['txn_mth'] = df_HM_CustTrxn['t_dat'].dt.strftime('%m')
df_HM_CustTrxn['txn_mth'] = df_HM_CustTrxn['txn_mth'].astype(str).astype(int)

# Add Column for Transaction Day and convert to integer value
df_HM_CustTrxn['txn_day'] = df_HM_CustTrxn['t_dat'].dt.strftime('%d')
df_HM_CustTrxn['txn_day'] = df_HM_CustTrxn['txn_day'].astype(str).astype(int)

In [233]:
# Filter out column to drop timestamp attribute
df_HM_CustTrxn = df_HM_CustTrxn.iloc[:,1:8]

In [234]:
df_HM_CustTrxn.head()

customer_id  article_id  \
4413783   24df7e5a17b32fb47e413c0909ea38794ebb3b5f48987f...   706016002   
15953688  2a009e5b897a07be7a7c22924c41dc35ae2706363163fc...   721932002   
18004153  42d005d022b0afee473b01ab3ab3fc6ccfea377b520e55...   688558002   
15948120  fb8fc6d3f2f59376d4e7d79273e3aa3278fa5472edc5bf...   331093006   
18584603  4c598e57d6c33608afff9fd0f76a94b4964fc907bd69ee...   746775008   

             price  sales_channel_id  txn_year  txn_mth  txn_day  
4413783   0.033881                 2      2019        1        1  
15953688  0.045746                 2      2019        8       27  
18004153  0.016932                 1      2019       10       17  
15948120  0.025407                 2      2019        8       26  
18584603  0.020322                 2      2019       11        3

In [235]:
df_HM_CustTrxn.dtypes

customer_id          object
article_id            int64
price               float64
sales_channel_id      int64
txn_year              int32
txn_mth               int32
txn_day               int32
dtype: object

In [236]:
# Check for null values in the Transaction Dataframe
[print('Any Null Values in the Trxn (sample) dataframe?','\n' ,df_HM_CustTrxn.isnull().any())]

Any Null Values in the Trxn (sample) dataframe? 
 customer_id         False
article_id          False
price               False
sales_channel_id    False
txn_year            False
txn_mth             False
txn_day             False
dtype: bool


[None]

#### Extend Dataframe with Customer Data

In [237]:
# Confirm dimensions of Trxn (sample) dataframe BEFORE merging with Customer data
df_HM_CustTrxn.shape

(31788, 7)

#### Merge Customer Attributes into the Trxn (Sample) Dataframe. <a class="tocSkip">

In [238]:
# Merge Customer atributes from the Customer dataset based on a join with the 'customer_id' field
df_HM_CustTrxn = pd.merge(df_HM_CustTrxn, df_customers_sample, how="left", on=["customer_id"])

In [239]:
# Confirm dimensions of Trxn (sample) dataframe AFTER merging with Customer data
df_HM_CustTrxn.shape

(31788, 13)

In [240]:
df_HM_CustTrxn.head()

customer_id  article_id     price  \
0  24df7e5a17b32fb47e413c0909ea38794ebb3b5f48987f...   706016002  0.033881   
1  2a009e5b897a07be7a7c22924c41dc35ae2706363163fc...   721932002  0.045746   
2  42d005d022b0afee473b01ab3ab3fc6ccfea377b520e55...   688558002  0.016932   
3  fb8fc6d3f2f59376d4e7d79273e3aa3278fa5472edc5bf...   331093006  0.025407   
4  4c598e57d6c33608afff9fd0f76a94b4964fc907bd69ee...   746775008  0.020322   

   sales_channel_id  txn_year  txn_mth  txn_day   FN  Active  \
0                 2      2019        1        1  1.0     1.0   
1                 2      2019        8       27  1.0     1.0   
2                 1      2019       10       17  NaN     NaN   
3                 2      2019        8       26  NaN     NaN   
4                 2      2019       11        3  1.0     1.0   

  club_member_status fashion_news_frequency   age  \
0             ACTIVE              Regularly  20.0   
1             ACTIVE              Regularly  27.0   
2             ACTIVE                    NaN  28.0   
3         PRE-CREATE                   NONE  29.0   
4             ACTIVE              Regularly  43.0   

                                         postal_code  
0  99b10967745ac020c48aaa87c2ebb5022dd3e3fbe7e5b6...  
1  5706de11e3bd87cdff219b03e8af086e8e29dcdc517d70...  
2  92ed066e76c70c60d32f1078588267b2e5145ca4193373...  
3  86e76c24f5609e780ee3abf6ab784956b02c6948923353...  
4  070b1212e53dde8ff26f55c9dfd5b482f35647ef3469fc...

In [241]:
# Filter out other Customer specific attributes apart from; age + fashion_news_frequency + club_member_status + postal code
# There are very many missing values in the other Customer dataset columns, so these will be ignored
df_HM_CustTrxn = df_HM_CustTrxn.iloc[:, [0, 1, 2, 3, 4, 5, 6, 9, 10, 11, 12]]

In [242]:
df_HM_CustTrxn.head()

customer_id  article_id     price  \
0  24df7e5a17b32fb47e413c0909ea38794ebb3b5f48987f...   706016002  0.033881   
1  2a009e5b897a07be7a7c22924c41dc35ae2706363163fc...   721932002  0.045746   
2  42d005d022b0afee473b01ab3ab3fc6ccfea377b520e55...   688558002  0.016932   
3  fb8fc6d3f2f59376d4e7d79273e3aa3278fa5472edc5bf...   331093006  0.025407   
4  4c598e57d6c33608afff9fd0f76a94b4964fc907bd69ee...   746775008  0.020322   

   sales_channel_id  txn_year  txn_mth  txn_day club_member_status  \
0                 2      2019        1        1             ACTIVE   
1                 2      2019        8       27             ACTIVE   
2                 1      2019       10       17             ACTIVE   
3                 2      2019        8       26         PRE-CREATE   
4                 2      2019       11        3             ACTIVE   

  fashion_news_frequency   age  \
0              Regularly  20.0   
1              Regularly  27.0   
2                    NaN  28.0   
3                   NONE  29.0   
4              Regularly  43.0   

                                         postal_code  
0  99b10967745ac020c48aaa87c2ebb5022dd3e3fbe7e5b6...  
1  5706de11e3bd87cdff219b03e8af086e8e29dcdc517d70...  
2  92ed066e76c70c60d32f1078588267b2e5145ca4193373...  
3  86e76c24f5609e780ee3abf6ab784956b02c6948923353...  
4  070b1212e53dde8ff26f55c9dfd5b482f35647ef3469fc...

In [243]:
df_HM_CustTrxn.shape

(31788, 11)

In [244]:
# Check for null values in the merged Trxn/Customer dataframe
[print('Any Null Values in the merged Trxn/Customer dataframe?','\n' ,df_HM_CustTrxn.isnull().any())]

Any Null Values in the merged Trxn/Customer dataframe? 
 customer_id               False
article_id                False
price                     False
sales_channel_id          False
txn_year                  False
txn_mth                   False
txn_day                   False
club_member_status         True
fashion_news_frequency     True
age                        True
postal_code               False
dtype: bool


[None]

#### Remove Rows with Missing Club Member Status Attributes. <a class="tocSkip">

In [245]:
# Check missing values in the club_member_status column  
df_null_club_member_status = df_HM_CustTrxn.loc[(df_HM_CustTrxn["club_member_status"].isnull())]

[print('Numbers of rows with missing club member status?','\n' ,len(df_null_club_member_status.index))]
[print('\nPrecentage of total rows in Trxn dataframe with missing club member status attributes?\n{:06.3f}%'
                               .format((len(df_null_club_member_status.index)/len(df_HM_CustTrxn.index))*100))]


# For simplicity the rows with missing values in the club member status will be ignored
# Re-Build Txn/Customer Dataframe with no missing values in the club member status
df_HM_CustTrxn = df_HM_CustTrxn.loc[(df_HM_CustTrxn["club_member_status"].notnull())]

# Confirm dimensions of Trxn (sample) dataframe AFTER removing rows with missing values in the club member status column
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - with missing club mem status values removed','\n' ,
                                                                                                   df_HM_CustTrxn.shape)]


Numbers of rows with missing club member status? 
 70

Precentage of total rows in Trxn dataframe with missing club member status attributes?
00.220%

Current Number of Rows/Columns in Trxn/Customer dataframe - with missing club mem status values removed 
 (31718, 11)


[None]

#### Remove Rows with Missing Fashion News Frequency Attributes. <a class="tocSkip">

In [246]:
# Check missing values in the club_member_status column  
df_null_fn_freq = df_HM_CustTrxn.loc[(df_HM_CustTrxn["fashion_news_frequency"].isnull())]

[print('Numbers of rows with missing fashion_news_frequency values?','\n' ,len(df_null_fn_freq.index))]
[print('\nPrecentage of total rows in Trxn dataframe with missing fashion_news_frequency attributes?\n{:06.3f}%'
                               .format((len(df_null_fn_freq.index)/len(df_HM_CustTrxn.index))*100))]

# For simplicity the rows with missing values in the fashion_news_frequency column will be ignored
# Re-Build Txn/Customer Dataframe with no missing values in the fashion_news_frequency column
df_HM_CustTrxn = df_HM_CustTrxn.loc[(df_HM_CustTrxn["fashion_news_frequency"].notnull())]

# Confirm dimensions of Trxn (sample) dataframe AFTER removing rows with missing values in the fashion_news_frequency column
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - missing Fash News Freq values removed','\n' ,
                                                                                                   df_HM_CustTrxn.shape)]

Numbers of rows with missing fashion_news_frequency values? 
 123

Precentage of total rows in Trxn dataframe with missing fashion_news_frequency attributes?
00.388%

Current Number of Rows/Columns in Trxn/Customer dataframe - missing Fash News Freq values removed 
 (31595, 11)


[None]

#### Remove Rows with Missing Age Attributes. <a class="tocSkip">

In [247]:
# Check missing values in the age column  
df_null_age = df_HM_CustTrxn.loc[(df_HM_CustTrxn["age"].isnull())]

[print('Numbers of rows with missing age values?','\n' ,len(df_null_age.index))]
[print('\nPrecentage of total rows in Trxn dataframe with missing age attributes?\n{:06.3f}%'
                               .format((len(df_null_age.index)/len(df_HM_CustTrxn.index))*100))]

# Re-Build Txn/Customer Dataframe with no missing values in the age column
df_HM_CustTrxn = df_HM_CustTrxn.loc[(df_HM_CustTrxn["age"].notnull())]


# Confirm dimensions of Trxn (sample) dataframe AFTER removing rows with missing values in the age column
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - missing ages removed','\n' ,df_HM_CustTrxn.shape)]

Numbers of rows with missing age values? 
 117

Precentage of total rows in Trxn dataframe with missing age attributes?
00.370%

Current Number of Rows/Columns in Trxn/Customer dataframe - missing ages removed 
 (31478, 11)


[None]

#### Encode the club_member_status + fashion_news_frequency attributes <a class="tocSkip">

In [248]:
# The categorical values for club_member_status must be encoded to numberical values
df_encoded_member_stats = pd.get_dummies(df_HM_CustTrxn['club_member_status'])

In [249]:
df_encoded_member_stats

ACTIVE  LEFT CLUB  PRE-CREATE
0           1          0           0
1           1          0           0
3           0          0           1
4           1          0           0
5           1          0           0
...       ...        ...         ...
31783       1          0           0
31784       1          0           0
31785       1          0           0
31786       1          0           0
31787       1          0           0

[31478 rows x 3 columns]

In [250]:
# Add Encoded club_member_status to our merged H&M dataframe
df_HM_CustTrxn = df_HM_CustTrxn.join(df_encoded_member_stats)

In [251]:
df_HM_CustTrxn.head()

customer_id  article_id     price  \
0  24df7e5a17b32fb47e413c0909ea38794ebb3b5f48987f...   706016002  0.033881   
1  2a009e5b897a07be7a7c22924c41dc35ae2706363163fc...   721932002  0.045746   
3  fb8fc6d3f2f59376d4e7d79273e3aa3278fa5472edc5bf...   331093006  0.025407   
4  4c598e57d6c33608afff9fd0f76a94b4964fc907bd69ee...   746775008  0.020322   
5  ba8ad8f6c299bc96d1fcdc412e9dd26fefebe74fb9c62d...   700753002  0.025407   

   sales_channel_id  txn_year  txn_mth  txn_day club_member_status  \
0                 2      2019        1        1             ACTIVE   
1                 2      2019        8       27             ACTIVE   
3                 2      2019        8       26         PRE-CREATE   
4                 2      2019       11        3             ACTIVE   
5                 2      2019        1       30             ACTIVE   

  fashion_news_frequency   age  \
0              Regularly  20.0   
1              Regularly  27.0   
3                   NONE  29.0   
4              Regularly  43.0   
5                   NONE  21.0   

                                         postal_code  ACTIVE  LEFT CLUB  \
0  99b10967745ac020c48aaa87c2ebb5022dd3e3fbe7e5b6...       1          0   
1  5706de11e3bd87cdff219b03e8af086e8e29dcdc517d70...       1          0   
3  86e76c24f5609e780ee3abf6ab784956b02c6948923353...       0          0   
4  070b1212e53dde8ff26f55c9dfd5b482f35647ef3469fc...       1          0   
5  523628d9b1fe526168d42333430fadca96338c40a39bf2...       1          0   

   PRE-CREATE  
0           0  
1           0  
3           1  
4           0  
5           0

In [252]:
# The categorical values for fashion_news_frequency must be encoded to numberical values
df_encoded_fn_freq = pd.get_dummies(df_HM_CustTrxn['fashion_news_frequency'])

In [253]:
df_encoded_fn_freq

Monthly  NONE  Regularly
0            0     0          1
1            0     0          1
3            0     1          0
4            0     0          1
5            0     1          0
...        ...   ...        ...
31783        0     1          0
31784        0     1          0
31785        0     1          0
31786        0     1          0
31787        0     1          0

[31478 rows x 3 columns]

In [254]:
# Add Encoded fashion_news_frequency to our merged H&M dataframe
df_HM_CustTrxn = df_HM_CustTrxn.join(df_encoded_fn_freq)

In [255]:
df_HM_CustTrxn.head()

customer_id  article_id     price  \
0  24df7e5a17b32fb47e413c0909ea38794ebb3b5f48987f...   706016002  0.033881   
1  2a009e5b897a07be7a7c22924c41dc35ae2706363163fc...   721932002  0.045746   
3  fb8fc6d3f2f59376d4e7d79273e3aa3278fa5472edc5bf...   331093006  0.025407   
4  4c598e57d6c33608afff9fd0f76a94b4964fc907bd69ee...   746775008  0.020322   
5  ba8ad8f6c299bc96d1fcdc412e9dd26fefebe74fb9c62d...   700753002  0.025407   

   sales_channel_id  txn_year  txn_mth  txn_day club_member_status  \
0                 2      2019        1        1             ACTIVE   
1                 2      2019        8       27             ACTIVE   
3                 2      2019        8       26         PRE-CREATE   
4                 2      2019       11        3             ACTIVE   
5                 2      2019        1       30             ACTIVE   

  fashion_news_frequency   age  \
0              Regularly  20.0   
1              Regularly  27.0   
3                   NONE  29.0   
4              Regularly  43.0   
5                   NONE  21.0   

                                         postal_code  ACTIVE  LEFT CLUB  \
0  99b10967745ac020c48aaa87c2ebb5022dd3e3fbe7e5b6...       1          0   
1  5706de11e3bd87cdff219b03e8af086e8e29dcdc517d70...       1          0   
3  86e76c24f5609e780ee3abf6ab784956b02c6948923353...       0          0   
4  070b1212e53dde8ff26f55c9dfd5b482f35647ef3469fc...       1          0   
5  523628d9b1fe526168d42333430fadca96338c40a39bf2...       1          0   

   PRE-CREATE  Monthly  NONE  Regularly  
0           0        0     0          1  
1           0        0     0          1  
3           1        0     1          0  
4           0        0     0          1  
5           0        0     1          0

In [256]:
# Remove the prior Categorical columns from the Trxn/Customer dataframe
# The 'inplace' parameter confirms that the drop operation takes place in the same dataframe (a new one is not created)
df_HM_CustTrxn.drop(["club_member_status","fashion_news_frequency"], axis=1, inplace=True)

In [257]:
# Drop Postal Code - a non numerical value that is not easily converted
df_HM_CustTrxn.drop(["postal_code"], axis=1, inplace=True)

In [258]:
df_HM_CustTrxn.head()

customer_id  article_id     price  \
0  24df7e5a17b32fb47e413c0909ea38794ebb3b5f48987f...   706016002  0.033881   
1  2a009e5b897a07be7a7c22924c41dc35ae2706363163fc...   721932002  0.045746   
3  fb8fc6d3f2f59376d4e7d79273e3aa3278fa5472edc5bf...   331093006  0.025407   
4  4c598e57d6c33608afff9fd0f76a94b4964fc907bd69ee...   746775008  0.020322   
5  ba8ad8f6c299bc96d1fcdc412e9dd26fefebe74fb9c62d...   700753002  0.025407   

   sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  LEFT CLUB  \
0                 2      2019        1        1  20.0       1          0   
1                 2      2019        8       27  27.0       1          0   
3                 2      2019        8       26  29.0       0          0   
4                 2      2019       11        3  43.0       1          0   
5                 2      2019        1       30  21.0       1          0   

   PRE-CREATE  Monthly  NONE  Regularly  
0           0        0     0          1  
1           0        0     0          1  
3           1        0     1          0  
4           0        0     0          1  
5           0        0     1          0

In [259]:
df_HM_CustTrxn.dtypes

customer_id          object
article_id            int64
price               float64
sales_channel_id      int64
txn_year              int32
txn_mth               int32
txn_day               int32
age                 float64
ACTIVE                uint8
LEFT CLUB             uint8
PRE-CREATE            uint8
Monthly               uint8
NONE                  uint8
Regularly             uint8
dtype: object

In [260]:
# Confirm dimensions of Trxn (sample) dataframe AFTER encoding Categorical Values
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - after encoding of categorical values completed','\n' ,
                                                                                                   df_HM_CustTrxn.shape)]


Current Number of Rows/Columns in Trxn/Customer dataframe - after encoding of categorical values completed 
 (31478, 14)


[None]

In [261]:
# Check for null values in the Transaction Dataframe
[print('Any Null Values in the merged Trxn/Customer (sample) dataframe?','\n' ,df_HM_CustTrxn.isnull().any())]

Any Null Values in the merged Trxn/Customer (sample) dataframe? 
 customer_id         False
article_id          False
price               False
sales_channel_id    False
txn_year            False
txn_mth             False
txn_day             False
age                 False
ACTIVE              False
LEFT CLUB           False
PRE-CREATE          False
Monthly             False
NONE                False
Regularly           False
dtype: bool


[None]

#### Extend Dataframe with Articles Data

#### Remove 'Redundant' Categorical Columns in Articles (sample) Dataframe. <a class="tocSkip">

In [262]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code                  prod_name  product_type_no  \
0   108775015        108775                  Strap top              253   
1   108775044        108775                  Strap top              253   
6   111565001        111565        20 den 1p Stockings              304   
8   111586001        111586  Shape Up 30 den 1p Tights              273   
9   111593001        111593   Support 40 den 1p Tights              304   

  product_type_name  product_group_name  graphical_appearance_no  \
0          Vest top  Garment Upper body                  1010016   
1          Vest top  Garment Upper body                  1010016   
6  Underwear Tights      Socks & Tights                  1010016   
8   Leggings/Tights  Garment Lower body                  1010016   
9  Underwear Tights      Socks & Tights                  1010016   

  graphical_appearance_name  colour_group_code colour_group_name  ...  \
0                     Solid                  9             Black  ...   
1                     Solid                 10             White  ...   
6                     Solid                  9             Black  ...   
8                     Solid                  9             Black  ...   
9                     Solid                  9             Black  ...   

   department_name index_code        index_name index_group_no  \
0     Jersey Basic          A        Ladieswear              1   
1     Jersey Basic          A        Ladieswear              1   
6     Tights basic          B  Lingeries/Tights              1   
8     Tights basic          B  Lingeries/Tights              1   
9     Tights basic          B  Lingeries/Tights              1   

   index_group_name section_no                    section_name  \
0        Ladieswear         16          Womens Everyday Basics   
1        Ladieswear         16          Womens Everyday Basics   
6        Ladieswear         62  Womens Nightwear, Socks & Tigh   
8        Ladieswear         62  Womens Nightwear, Socks & Tigh   
9        Ladieswear         62  Womens Nightwear, Socks & Tigh   

  garment_group_no  garment_group_name  \
0             1002        Jersey Basic   
1             1002        Jersey Basic   
6             1021    Socks and Tights   
8             1021    Socks and Tights   
9             1021    Socks and Tights   

                                         detail_desc  
0            Jersey top with narrow shoulder straps.  
1            Jersey top with narrow shoulder straps.  
6  Semi shiny nylon stockings with a wide, reinfo...  
8  Tights with built-in support to lift the botto...  
9  Semi shiny tights that shape the tummy, thighs...  

[5 rows x 25 columns]

In [263]:
# Remove 'redundant' categotical columns from Articles (sample) dataframe
# Consider a categorical column Redundant when is 'partnered' with a numerical code 
df_articles_sample.drop(["prod_name", 
                         "product_type_name", 
                         "graphical_appearance_name",
                         "colour_group_name",
                         "perceived_colour_value_name",
                         "perceived_colour_master_name",
                         "department_name",
                         "index_name",
                         "index_group_name",
                         "section_name",
                         "garment_group_name",
                         "detail_desc"], axis=1, inplace=True)

In [264]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code  product_type_no  product_group_name  \
0   108775015        108775              253  Garment Upper body   
1   108775044        108775              253  Garment Upper body   
6   111565001        111565              304      Socks & Tights   
8   111586001        111586              273  Garment Lower body   
9   111593001        111593              304      Socks & Tights   

   graphical_appearance_no  colour_group_code  perceived_colour_value_id  \
0                  1010016                  9                          4   
1                  1010016                 10                          3   
6                  1010016                  9                          4   
8                  1010016                  9                          4   
9                  1010016                  9                          4   

   perceived_colour_master_id  department_no index_code  index_group_no  \
0                           5           1676          A               1   
1                           9           1676          A               1   
6                           5           3608          B               1   
8                           5           3608          B               1   
9                           5           3608          B               1   

   section_no  garment_group_no  
0          16              1002  
1          16              1002  
6          62              1021  
8          62              1021  
9          62              1021

#### Use One-Hot-Encoding on remaining categorical columns. <a class="tocSkip">

In [265]:
# The categorical values for product_group_name + index_code must be encoded to numberical values
df_encoded_prd_grp_nm = pd.get_dummies(df_articles_sample['product_group_name'])
df_encoded_idx_cd     = pd.get_dummies(df_articles_sample['index_code'])

In [266]:
df_encoded_prd_grp_nm

Accessories  Bags  Cosmetic  Furniture  Garment Full body  \
0                 0     0         0          0                  0   
1                 0     0         0          0                  0   
6                 0     0         0          0                  0   
8                 0     0         0          0                  0   
9                 0     0         0          0                  0   
...             ...   ...       ...        ...                ...   
105351            0     0         0          0                  0   
105465            0     0         0          0                  0   
105485            0     0         0          0                  0   
105506            0     0         0          0                  0   
105510            0     0         0          0                  0   

        Garment Lower body  Garment Upper body  Garment and Shoe care  Items  \
0                        0                   1                      0      0   
1                        0                   1                      0      0   
6                        0                   0                      0      0   
8                        1                   0                      0      0   
9                        0                   0                      0      0   
...                    ...                 ...                    ...    ...   
105351                   0                   1                      0      0   
105465                   0                   0                      0      0   
105485                   0                   0                      0      0   
105506                   0                   1                      0      0   
105510                   1                   0                      0      0   

        Nightwear  Shoes  Socks & Tights  Swimwear  Underwear  Unknown  
0               0      0               0         0          0        0  
1               0      0               0         0          0        0  
6               0      0               1         0          0        0  
8               0      0               0         0          0        0  
9               0      0               1         0          0        0  
...           ...    ...             ...       ...        ...      ...  
105351          0      0               0         0          0        0  
105465          0      0               1         0          0        0  
105485          0      0               0         0          1        0  
105506          0      0               0         0          0        0  
105510          0      0               0         0          0        0  

[18361 rows x 15 columns]

In [267]:
df_encoded_idx_cd

A  B  C  D  F  G  H  I  J  S
0       1  0  0  0  0  0  0  0  0  0
1       1  0  0  0  0  0  0  0  0  0
6       0  1  0  0  0  0  0  0  0  0
8       0  1  0  0  0  0  0  0  0  0
9       0  1  0  0  0  0  0  0  0  0
...    .. .. .. .. .. .. .. .. .. ..
105351  1  0  0  0  0  0  0  0  0  0
105465  0  1  0  0  0  0  0  0  0  0
105485  1  0  0  0  0  0  0  0  0  0
105506  1  0  0  0  0  0  0  0  0  0
105510  1  0  0  0  0  0  0  0  0  0

[18361 rows x 10 columns]

In [268]:
# Add Encoded product_group_name to our articles dataframe (sample)
df_articles_sample = df_articles_sample.join(df_encoded_prd_grp_nm)
df_articles_sample = df_articles_sample.join(df_encoded_idx_cd)

In [269]:
df_articles_sample.shape

(18361, 38)

In [270]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code  product_type_no  product_group_name  \
0   108775015        108775              253  Garment Upper body   
1   108775044        108775              253  Garment Upper body   
6   111565001        111565              304      Socks & Tights   
8   111586001        111586              273  Garment Lower body   
9   111593001        111593              304      Socks & Tights   

   graphical_appearance_no  colour_group_code  perceived_colour_value_id  \
0                  1010016                  9                          4   
1                  1010016                 10                          3   
6                  1010016                  9                          4   
8                  1010016                  9                          4   
9                  1010016                  9                          4   

   perceived_colour_master_id  department_no index_code  ...  A  B  C  D  F  \
0                           5           1676          A  ...  1  0  0  0  0   
1                           9           1676          A  ...  1  0  0  0  0   
6                           5           3608          B  ...  0  1  0  0  0   
8                           5           3608          B  ...  0  1  0  0  0   
9                           5           3608          B  ...  0  1  0  0  0   

   G  H  I  J  S  
0  0  0  0  0  0  
1  0  0  0  0  0  
6  0  0  0  0  0  
8  0  0  0  0  0  
9  0  0  0  0  0  

[5 rows x 38 columns]

In [271]:
# Remove the prior Categorical columns from the Articles dataframe
df_articles_sample.drop(["product_group_name","index_code"], axis=1, inplace=True)

In [272]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code  product_type_no  graphical_appearance_no  \
0   108775015        108775              253                  1010016   
1   108775044        108775              253                  1010016   
6   111565001        111565              304                  1010016   
8   111586001        111586              273                  1010016   
9   111593001        111593              304                  1010016   

   colour_group_code  perceived_colour_value_id  perceived_colour_master_id  \
0                  9                          4                           5   
1                 10                          3                           9   
6                  9                          4                           5   
8                  9                          4                           5   
9                  9                          4                           5   

   department_no  index_group_no  section_no  ...  A  B  C  D  F  G  H  I  J  \
0           1676               1          16  ...  1  0  0  0  0  0  0  0  0   
1           1676               1          16  ...  1  0  0  0  0  0  0  0  0   
6           3608               1          62  ...  0  1  0  0  0  0  0  0  0   
8           3608               1          62  ...  0  1  0  0  0  0  0  0  0   
9           3608               1          62  ...  0  1  0  0  0  0  0  0  0   

   S  
0  0  
1  0  
6  0  
8  0  
9  0  

[5 rows x 36 columns]

In [273]:
df_articles_sample.dtypes

article_id                    int64
product_code                  int64
product_type_no               int64
graphical_appearance_no       int64
colour_group_code             int64
perceived_colour_value_id     int64
perceived_colour_master_id    int64
department_no                 int64
index_group_no                int64
section_no                    int64
garment_group_no              int64
Accessories                   uint8
Bags                          uint8
Cosmetic                      uint8
Furniture                     uint8
Garment Full body             uint8
Garment Lower body            uint8
Garment Upper body            uint8
Garment and Shoe care         uint8
Items                         uint8
Nightwear                     uint8
Shoes                         uint8
Socks & Tights                uint8
Swimwear                      uint8
Underwear                     uint8
Unknown                       uint8
A                             uint8
B                           

#### Merge Trxn/Customers Dataframe with Articles Dataframe (sample). <a class="tocSkip">

In [274]:
# Merge Articles atributes from the Articles dataset based on a join with the 'article_id' field
df_HM_CustTrxnArtcls = pd.merge(df_HM_CustTrxn, df_articles_sample, how="left", on=["article_id"])

In [275]:
# Confirm dimensions of Trxn (sample) dataframe AFTER merging with Articles data
df_HM_CustTrxnArtcls.shape

(31478, 49)

In [276]:
# Display basic column information for the initial Kaggle 'transaction_train' datatset - AFTER merging with Articles data
df_HM_CustTrxnArtcls.head()

customer_id  article_id     price  \
0  24df7e5a17b32fb47e413c0909ea38794ebb3b5f48987f...   706016002  0.033881   
1  2a009e5b897a07be7a7c22924c41dc35ae2706363163fc...   721932002  0.045746   
2  fb8fc6d3f2f59376d4e7d79273e3aa3278fa5472edc5bf...   331093006  0.025407   
3  4c598e57d6c33608afff9fd0f76a94b4964fc907bd69ee...   746775008  0.020322   
4  ba8ad8f6c299bc96d1fcdc412e9dd26fefebe74fb9c62d...   700753002  0.025407   

   sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  LEFT CLUB  ...  \
0                 2      2019        1        1  20.0       1          0  ...   
1                 2      2019        8       27  27.0       1          0  ...   
2                 2      2019        8       26  29.0       0          0  ...   
3                 2      2019       11        3  43.0       1          0  ...   
4                 2      2019        1       30  21.0       1          0  ...   

   A  B  C  D  F  G  H  I  J  S  
0  0  0  0  1  0  0  0  0  0  0  
1  1  0  0  0  0  0  0  0  0  0  
2  0  0  0  1  0  0  0  0  0  0  
3  0  1  0  0  0  0  0  0  0  0  
4  0  1  0  0  0  0  0  0  0  0  

[5 rows x 49 columns]

In [277]:
df_HM_CustTrxnArtcls.dtypes

customer_id                    object
article_id                      int64
price                         float64
sales_channel_id                int64
txn_year                        int32
txn_mth                         int32
txn_day                         int32
age                           float64
ACTIVE                          uint8
LEFT CLUB                       uint8
PRE-CREATE                      uint8
Monthly                         uint8
NONE                            uint8
Regularly                       uint8
product_code                    int64
product_type_no                 int64
graphical_appearance_no         int64
colour_group_code               int64
perceived_colour_value_id       int64
perceived_colour_master_id      int64
department_no                   int64
index_group_no                  int64
section_no                      int64
garment_group_no                int64
Accessories                     uint8
Bags                            uint8
Cosmetic    

In [278]:
# Check for null values in the merged Trxn/Customer/Articles dataframe
[print('Any Null Values in the merged Trxn/Customer/Articles dataframe?','\n' ,df_HM_CustTrxnArtcls.isnull().any())]

Any Null Values in the merged Trxn/Customer/Articles dataframe? 
 customer_id                   False
article_id                    False
price                         False
sales_channel_id              False
txn_year                      False
txn_mth                       False
txn_day                       False
age                           False
ACTIVE                        False
LEFT CLUB                     False
PRE-CREATE                    False
Monthly                       False
NONE                          False
Regularly                     False
product_code                  False
product_type_no               False
graphical_appearance_no       False
colour_group_code             False
perceived_colour_value_id     False
perceived_colour_master_id    False
department_no                 False
index_group_no                False
section_no                    False
garment_group_no              False
Accessories                   False
Bags                          Fals

[None]

#### Final Data Preparation Steps

#### Remove Any Duplicates in Final Merged Dataframe. <a class="tocSkip">

In [279]:
# run a check to see if any duplicate entires exist in the merged Trxn/Customer/Articles dataframe
[print('Numbers of duplicated rows in the final merged Trxn/Customer/Articles dataframe?','\n' 
                                                                       ,df_HM_CustTrxnArtcls.duplicated().sum())]

Numbers of duplicated rows in the final merged Trxn/Customer/Articles dataframe? 
 2


[None]

In [280]:
# Find duplicate rows across all columns
duplicateRows = df_HM_CustTrxnArtcls[df_HM_CustTrxnArtcls.duplicated()]

In [281]:
duplicateRows

customer_id  article_id  \
18874  b3ebc67c50618fa84c60390a604cae10066da537d5bdc6...   297791001   
21354  216e9b46eabedc803df4beb1bb0c8662b44772f6f2e6f8...   572797001   

          price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  \
18874  0.012203                 2      2018       11       23  52.0       1   
21354  0.011847                 2      2020        3        3  24.0       1   

       LEFT CLUB  ...  A  B  C  D  F  G  H  I  J  S  
18874          0  ...  0  1  0  0  0  0  0  0  0  0  
21354          0  ...  1  0  0  0  0  0  0  0  0  0  

[2 rows x 49 columns]

In [282]:
# Drop duplicate rows in the merged Trxn/Customer/Articles dataframe
df_HM_CustTrxnArtcls = df_HM_CustTrxnArtcls.drop_duplicates()

In [283]:
# Re-check: Find duplicate rows across all columns
duplicateRows = df_HM_CustTrxnArtcls[df_HM_CustTrxnArtcls.duplicated()]

In [284]:
duplicateRows

Empty DataFrame
Columns: [customer_id, article_id, price, sales_channel_id, txn_year, txn_mth, txn_day, age, ACTIVE, LEFT CLUB, PRE-CREATE, Monthly, NONE, Regularly, product_code, product_type_no, graphical_appearance_no, colour_group_code, perceived_colour_value_id, perceived_colour_master_id, department_no, index_group_no, section_no, garment_group_no, Accessories, Bags, Cosmetic, Furniture, Garment Full body, Garment Lower body, Garment Upper body, Garment and Shoe care, Items, Nightwear, Shoes, Socks & Tights, Swimwear, Underwear, Unknown, A, B, C, D, F, G, H, I, J, S]
Index: []

[0 rows x 49 columns]

In [285]:
# Confirm dimensions of Trxn/Customer/Articles (sample) dataframe AFTER removing duplicate rows
[print('\nCurrent Number of Rows/Columns in merged Trxn/Customer/Articles dataframe - duplicates removed','\n' ,
                                                                                           df_HM_CustTrxnArtcls.shape)]


Current Number of Rows/Columns in merged Trxn/Customer/Articles dataframe - duplicates removed 
 (31476, 49)


[None]

In [286]:
### See if there is a way to check the integrity of the customer merging..

###### 

## Split Data: Train and Test Sets

#### Split Data into 70% for Training, 30% for Testing. <a class="tocSkip">

In [287]:
# Split the data into train and test set
# Parameter set to 30% for size of Test set
# Given that there is ordering in the dataset, the 'shuffle' parameter is set to 'True'
train, test = train_test_split(df_HM_CustTrxnArtcls, test_size=0.3, random_state=101, shuffle=True)

In [288]:
# Output Confirmation of data split
dataDescription = 'H&M Sample Data Split'
print("\n{} Training Set Shape : \n\t".format(dataDescription))
print(train.shape)
print("\n{} Test Set Shape : \n\t".format(dataDescription))
print(test.shape)


H&M Sample Data Split Training Set Shape : 
	
(22033, 49)

H&M Sample Data Split Test Set Shape : 
	
(9443, 49)


###### 

## Model Preparation

#### Scale the Training Dataset (but not the Test set) <a class="tocSkip">

In [289]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

In [290]:
train.head()

customer_id  article_id  \
20327  6793980a0733db0db2b37ee0d63dcdbca67d68e951a359...   693322006   
23850  2c4096d75492f8150cd724468d82dfad4deb74c82ebdff...   803757001   
9130   a406b87b55ec5be6dceb89de10a668719637d44e2a8971...   611522006   
11641  46ef71557837f56d4655cef3c5be315a8496886ec72512...   657510001   
30778  54cf679d5f790d76991d110d688d3ec40bd6b049edf80b...   316441005   

          price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  \
20327  0.019864                 2      2019        6       24  29.0       1   
23850  0.016932                 2      2020        2        2  27.0       1   
9130   0.022017                 2      2018        9       21  42.0       1   
11641  0.030492                 1      2018       11        4  27.0       1   
30778  0.016932                 1      2018       10       26  30.0       1   

       LEFT CLUB  ...  A  B  C  D  F  G  H  I  J  S  
20327          0  ...  0  0  0  0  0  0  0  0  0  1  
23850          0  ...  1  0  0  0  0  0  0  0  0  0  
9130           0  ...  0  0  0  0  0  0  0  1  0  0  
11641          0  ...  0  0  0  1  0  0  0  0  0  0  
30778          0  ...  1  0  0  0  0  0  0  0  0  0  

[5 rows x 49 columns]

In [291]:
chk_cols =  ['product_code',
            'product_type_no',
            'graphical_appearance_no',
            'colour_group_code',
            'perceived_colour_value_id',
            'perceived_colour_master_id',
            'department_no',
            'index_group_no',
            'section_no',
            'garment_group_no']

In [292]:
train[chk_cols]

product_code  product_type_no  graphical_appearance_no  \
20327        693322              274                  1010010   
23850        803757              272                  1010016   
9130         611522              272                  1010026   
11641        657510              252                  1010016   
30778        316441              254                  1010016   
...             ...              ...                      ...   
5695         663851              259                  1010004   
8006         724906              272                  1010023   
17745        712216              253                  1010016   
17931        719201              275                  1010023   
13151        730609              255                  1010005   

       colour_group_code  perceived_colour_value_id  \
20327                  7                          2   
23850                  9                          4   
9130                   9                          4   
11641                  9                          4   
30778                 10                          3   
...                  ...                        ...   
5695                   1                          6   
8006                  71                          1   
17745                 10                          3   
17931                 10                          3   
13151                 10                          3   

       perceived_colour_master_id  department_no  index_group_no  section_no  \
20327                          12           8310              26           5   
23850                           5           1676               1          16   
9130                            5           7930               4          79   
11641                           5           1640               2          53   
30778                           9           1676               1          16   
...                           ...            ...             ...         ...   
5695                           14           7657               4          46   
8006                            2           1772               2          57   
17745                           9           1522               1          15   
17931                           9           3030               1          15   
13151                           9           5848               3          21   

       garment_group_no  
20327              1005  
23850              1002  
9130               1002  
11641              1005  
30778              1002  
...                 ...  
5695               1011  
8006               1016  
17745              1010  
17931              1001  
13151              1005  

[22033 rows x 10 columns]

In [293]:
cols_to_norm = ['price','age']

In [294]:
#train[cols_to_norm] = scaler.fit_transform(train[cols_to_norm])

In [295]:
# Normalizing numerical features so that each feature has mean 0 and variance 1
feature_scaler = StandardScaler()
#train[cols_to_norm] = feature_scaler.fit_transform(train[cols_to_norm])

In [296]:
#for x in dataset.columns[dataset.dtypes == 'int64']:
#    Dataset[x] = minmax.fit_transform(np.array(Dataset[I]).reshape(-1,1))

In [297]:
def scale_trainset_features(df, dtype, no_scale):
    for x in df.columns[df.dtypes == dtype]:
        if x != no_scale:
            df[x] = scaler.fit_transform(np.array(df[x]).reshape(-1,1))    

In [298]:
#for x in train.columns[train.dtypes == 'int64']:
#    if x != 'article_id':
#        train[x] = scaler.fit_transform(np.array(train[x]).reshape(-1,1))

In [299]:
no_scale = 'article_id'
scale_dtype = 'int64'
scale_trainset_features(train, scale_dtype, no_scale)

In [300]:
train.head()

customer_id  article_id  \
20327  6793980a0733db0db2b37ee0d63dcdbca67d68e951a359...   693322006   
23850  2c4096d75492f8150cd724468d82dfad4deb74c82ebdff...   803757001   
9130   a406b87b55ec5be6dceb89de10a668719637d44e2a8971...   611522006   
11641  46ef71557837f56d4655cef3c5be315a8496886ec72512...   657510001   
30778  54cf679d5f790d76991d110d688d3ec40bd6b049edf80b...   316441005   

          price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  \
20327  0.019864               1.0      2019        6       24  29.0       1   
23850  0.016932               1.0      2020        2        2  27.0       1   
9130   0.022017               1.0      2018        9       21  42.0       1   
11641  0.030492               0.0      2018       11        4  27.0       1   
30778  0.016932               0.0      2018       10       26  30.0       1   

       LEFT CLUB  ...  A  B  C  D  F  G  H  I  J  S  
20327          0  ...  0  0  0  0  0  0  0  0  0  1  
23850          0  ...  1  0  0  0  0  0  0  0  0  0  
9130           0  ...  0  0  0  0  0  0  0  1  0  0  
11641          0  ...  0  0  0  1  0  0  0  0  0  0  
30778          0  ...  1  0  0  0  0  0  0  0  0  0  

[5 rows x 49 columns]

In [348]:
train[chk_cols]

customer_id  article_id  \
0      6793980a0733db0db2b37ee0d63dcdbca67d68e951a359...   693322006   
1      2c4096d75492f8150cd724468d82dfad4deb74c82ebdff...   803757001   
2      a406b87b55ec5be6dceb89de10a668719637d44e2a8971...   611522006   
3      46ef71557837f56d4655cef3c5be315a8496886ec72512...   657510001   
4      54cf679d5f790d76991d110d688d3ec40bd6b049edf80b...   316441005   
...                                                  ...         ...   
22028  330268eadb6b1b1a95876b6aab3e8b6b94fdafdd75c1bf...   663851002   
22029  f89ef8e2b20fd96ec23a89bfca01a04c1aae5b9aa86830...   724906003   
22030  45977583c4d69a0d5c843d96a7d375a75dc37af88c9864...   712216005   
22031  f25575d1bb6daf5835926aa5b06c9dc55e54ee576a8298...   719201002   
22032  41ea93e75936cf7b18251d420037216d38ec868b7ea34d...   730609001   

       product_code  product_type_no  graphical_appearance_no  \
0          0.697507         0.515947                 0.999982   
1          0.829283         0.512195                 0.999988   
2          0.599900         0.512195                 0.999998   
3          0.654774         0.474672                 0.999988   
4          0.247796         0.478424                 0.999988   
...             ...              ...                      ...   
22028      0.662341         0.487805                 0.999976   
22029      0.735194         0.512195                 0.999995   
22030      0.720052         0.476548                 0.999988   
22031      0.728387         0.517824                 0.999995   
22032      0.741999         0.480300                 0.999977   

       colour_group_code  perceived_colour_value_id  \
0               0.085106                      0.375   
1               0.106383                      0.625   
2               0.106383                      0.625   
3               0.106383                      0.625   
4               0.117021                      0.500   
...                  ...                        ...   
22028           0.021277                      0.875   
22029           0.765957                      0.250   
22030           0.117021                      0.500   
22031           0.117021                      0.500   
22032           0.117021                      0.500   

       perceived_colour_master_id  department_no  index_group_no  section_no  \
0                        0.619048       0.808944            1.00    0.031579   
1                        0.285714       0.054051            0.00    0.147368   
2                        0.285714       0.765703            0.12    0.810526   
3                        0.285714       0.049954            0.04    0.536842   
4                        0.476190       0.054051            0.00    0.147368   
...                           ...            ...             ...         ...   
22028                    0.714286       0.734638            0.12    0.463158   
22029                    0.142857       0.064975            0.04    0.578947   
22030                    0.476190       0.036527            0.00    0.136842   
22031                    0.476190       0.208125            0.00    0.136842   
22032                    0.476190       0.528789            0.08    0.200000   

       garment_group_no  
0              0.166667  
1              0.041667  
2              0.041667  
3              0.166667  
4              0.041667  
...                 ...  
22028          0.416667  
22029          0.625000  
22030          0.375000  
22031          0.000000  
22032          0.166667  

[22033 rows x 12 columns]

In [302]:
train.head()

customer_id  article_id  \
20327  6793980a0733db0db2b37ee0d63dcdbca67d68e951a359...   693322006   
23850  2c4096d75492f8150cd724468d82dfad4deb74c82ebdff...   803757001   
9130   a406b87b55ec5be6dceb89de10a668719637d44e2a8971...   611522006   
11641  46ef71557837f56d4655cef3c5be315a8496886ec72512...   657510001   
30778  54cf679d5f790d76991d110d688d3ec40bd6b049edf80b...   316441005   

          price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  \
20327  0.019864               1.0      2019        6       24  29.0       1   
23850  0.016932               1.0      2020        2        2  27.0       1   
9130   0.022017               1.0      2018        9       21  42.0       1   
11641  0.030492               0.0      2018       11        4  27.0       1   
30778  0.016932               0.0      2018       10       26  30.0       1   

       LEFT CLUB  ...  A  B  C  D  F  G  H  I  J  S  
20327          0  ...  0  0  0  0  0  0  0  0  0  1  
23850          0  ...  1  0  0  0  0  0  0  0  0  0  
9130           0  ...  0  0  0  0  0  0  0  1  0  0  
11641          0  ...  0  0  0  1  0  0  0  0  0  0  
30778          0  ...  1  0  0  0  0  0  0  0  0  0  

[5 rows x 49 columns]

In [303]:
#int32
#age                           float64

In [304]:
scale_dtype = 'int32'
scale_trainset_features(train, scale_dtype, no_scale)

In [305]:
scale_dtype = 'float64'
scale_trainset_features(train, scale_dtype, no_scale)

In [306]:
train.head() 

customer_id  article_id  \
20327  6793980a0733db0db2b37ee0d63dcdbca67d68e951a359...   693322006   
23850  2c4096d75492f8150cd724468d82dfad4deb74c82ebdff...   803757001   
9130   a406b87b55ec5be6dceb89de10a668719637d44e2a8971...   611522006   
11641  46ef71557837f56d4655cef3c5be315a8496886ec72512...   657510001   
30778  54cf679d5f790d76991d110d688d3ec40bd6b049edf80b...   316441005   

          price  sales_channel_id  txn_year   txn_mth   txn_day       age  \
20327  0.064430               1.0       0.5  0.454545  0.766667  0.171053   
23850  0.054754               1.0       1.0  0.090909  0.033333  0.144737   
9130   0.071532               1.0       0.0  0.727273  0.666667  0.342105   
11641  0.099497               0.0       0.0  0.909091  0.100000  0.144737   
30778  0.054754               0.0       0.0  0.818182  0.833333  0.184211   

       ACTIVE  LEFT CLUB  ...  A  B  C  D  F  G  H  I  J  S  
20327       1          0  ...  0  0  0  0  0  0  0  0  0  1  
23850       1          0  ...  1  0  0  0  0  0  0  0  0  0  
9130        1          0  ...  0  0  0  0  0  0  0  1  0  0  
11641       1          0  ...  0  0  0  1  0  0  0  0  0  0  
30778       1          0  ...  1  0  0  0  0  0  0  0  0  0  

[5 rows x 49 columns]

###### 

###### 

###### 

###### 

## K Nearest Neighbor Recommendations

In [307]:
from sklearn.neighbors import NearestNeighbors

In [308]:
model_knn = NearestNeighbors(metric = 'cosine', algorithm = 'brute')

In [309]:
train

customer_id  article_id  \
20327  6793980a0733db0db2b37ee0d63dcdbca67d68e951a359...   693322006   
23850  2c4096d75492f8150cd724468d82dfad4deb74c82ebdff...   803757001   
9130   a406b87b55ec5be6dceb89de10a668719637d44e2a8971...   611522006   
11641  46ef71557837f56d4655cef3c5be315a8496886ec72512...   657510001   
30778  54cf679d5f790d76991d110d688d3ec40bd6b049edf80b...   316441005   
...                                                  ...         ...   
5695   330268eadb6b1b1a95876b6aab3e8b6b94fdafdd75c1bf...   663851002   
8006   f89ef8e2b20fd96ec23a89bfca01a04c1aae5b9aa86830...   724906003   
17745  45977583c4d69a0d5c843d96a7d375a75dc37af88c9864...   712216005   
17931  f25575d1bb6daf5835926aa5b06c9dc55e54ee576a8298...   719201002   
13151  41ea93e75936cf7b18251d420037216d38ec868b7ea34d...   730609001   

          price  sales_channel_id  txn_year   txn_mth   txn_day       age  \
20327  0.064430               1.0       0.5  0.454545  0.766667  0.171053   
23850  0.054754               1.0       1.0  0.090909  0.033333  0.144737   
9130   0.071532               1.0       0.0  0.727273  0.666667  0.342105   
11641  0.099497               0.0       0.0  0.909091  0.100000  0.144737   
30778  0.054754               0.0       0.0  0.818182  0.833333  0.184211   
...         ...               ...       ...       ...       ...       ...   
5695   0.065940               0.0       0.0  0.818182  0.333333  0.342105   
8006   0.065940               1.0       1.0  0.181818  0.400000  0.407895   
17745  0.071532               1.0       0.5  0.636364  0.366667  0.210526   
17931  0.166611               1.0       0.5  0.272727  0.833333  0.394737   
13151  0.032383               1.0       0.5  0.545455  0.400000  0.184211   

       ACTIVE  LEFT CLUB  ...  A  B  C  D  F  G  H  I  J  S  
20327       1          0  ...  0  0  0  0  0  0  0  0  0  1  
23850       1          0  ...  1  0  0  0  0  0  0  0  0  0  
9130        1          0  ...  0  0  0  0  0  0  0  1  0  0  
11641       1          0  ...  0  0  0  1  0  0  0  0  0  0  
30778       1          0  ...  1  0  0  0  0  0  0  0  0  0  
...       ...        ...  ... .. .. .. .. .. .. .. .. .. ..  
5695        1          0  ...  0  0  0  0  0  0  1  0  0  0  
8006        1          0  ...  0  0  0  1  0  0  0  0  0  0  
17745       1          0  ...  1  0  0  0  0  0  0  0  0  0  
17931       1          0  ...  1  0  0  0  0  0  0  0  0  0  
13151       1          0  ...  0  0  0  0  1  0  0  0  0  0  

[22033 rows x 49 columns]

In [310]:
train.reset_index(drop=True,inplace=True)

In [311]:
train

customer_id  article_id  \
0      6793980a0733db0db2b37ee0d63dcdbca67d68e951a359...   693322006   
1      2c4096d75492f8150cd724468d82dfad4deb74c82ebdff...   803757001   
2      a406b87b55ec5be6dceb89de10a668719637d44e2a8971...   611522006   
3      46ef71557837f56d4655cef3c5be315a8496886ec72512...   657510001   
4      54cf679d5f790d76991d110d688d3ec40bd6b049edf80b...   316441005   
...                                                  ...         ...   
22028  330268eadb6b1b1a95876b6aab3e8b6b94fdafdd75c1bf...   663851002   
22029  f89ef8e2b20fd96ec23a89bfca01a04c1aae5b9aa86830...   724906003   
22030  45977583c4d69a0d5c843d96a7d375a75dc37af88c9864...   712216005   
22031  f25575d1bb6daf5835926aa5b06c9dc55e54ee576a8298...   719201002   
22032  41ea93e75936cf7b18251d420037216d38ec868b7ea34d...   730609001   

          price  sales_channel_id  txn_year   txn_mth   txn_day       age  \
0      0.064430               1.0       0.5  0.454545  0.766667  0.171053   
1      0.054754               1.0       1.0  0.090909  0.033333  0.144737   
2      0.071532               1.0       0.0  0.727273  0.666667  0.342105   
3      0.099497               0.0       0.0  0.909091  0.100000  0.144737   
4      0.054754               0.0       0.0  0.818182  0.833333  0.184211   
...         ...               ...       ...       ...       ...       ...   
22028  0.065940               0.0       0.0  0.818182  0.333333  0.342105   
22029  0.065940               1.0       1.0  0.181818  0.400000  0.407895   
22030  0.071532               1.0       0.5  0.636364  0.366667  0.210526   
22031  0.166611               1.0       0.5  0.272727  0.833333  0.394737   
22032  0.032383               1.0       0.5  0.545455  0.400000  0.184211   

       ACTIVE  LEFT CLUB  ...  A  B  C  D  F  G  H  I  J  S  
0           1          0  ...  0  0  0  0  0  0  0  0  0  1  
1           1          0  ...  1  0  0  0  0  0  0  0  0  0  
2           1          0  ...  0  0  0  0  0  0  0  1  0  0  
3           1          0  ...  0  0  0  1  0  0  0  0  0  0  
4           1          0  ...  1  0  0  0  0  0  0  0  0  0  
...       ...        ...  ... .. .. .. .. .. .. .. .. .. ..  
22028       1          0  ...  0  0  0  0  0  0  1  0  0  0  
22029       1          0  ...  0  0  0  1  0  0  0  0  0  0  
22030       1          0  ...  1  0  0  0  0  0  0  0  0  0  
22031       1          0  ...  1  0  0  0  0  0  0  0  0  0  
22032       1          0  ...  0  0  0  0  1  0  0  0  0  0  

[22033 rows x 49 columns]

In [312]:
df_cust_articles = train[chk_cols]

In [313]:
df_cust_articles

product_code  product_type_no  graphical_appearance_no  \
0          0.697507         0.515947                 0.999982   
1          0.829283         0.512195                 0.999988   
2          0.599900         0.512195                 0.999998   
3          0.654774         0.474672                 0.999988   
4          0.247796         0.478424                 0.999988   
...             ...              ...                      ...   
22028      0.662341         0.487805                 0.999976   
22029      0.735194         0.512195                 0.999995   
22030      0.720052         0.476548                 0.999988   
22031      0.728387         0.517824                 0.999995   
22032      0.741999         0.480300                 0.999977   

       colour_group_code  perceived_colour_value_id  \
0               0.085106                      0.375   
1               0.106383                      0.625   
2               0.106383                      0.625   
3               0.106383                      0.625   
4               0.117021                      0.500   
...                  ...                        ...   
22028           0.021277                      0.875   
22029           0.765957                      0.250   
22030           0.117021                      0.500   
22031           0.117021                      0.500   
22032           0.117021                      0.500   

       perceived_colour_master_id  department_no  index_group_no  section_no  \
0                        0.619048       0.808944            1.00    0.031579   
1                        0.285714       0.054051            0.00    0.147368   
2                        0.285714       0.765703            0.12    0.810526   
3                        0.285714       0.049954            0.04    0.536842   
4                        0.476190       0.054051            0.00    0.147368   
...                           ...            ...             ...         ...   
22028                    0.714286       0.734638            0.12    0.463158   
22029                    0.142857       0.064975            0.04    0.578947   
22030                    0.476190       0.036527            0.00    0.136842   
22031                    0.476190       0.208125            0.00    0.136842   
22032                    0.476190       0.528789            0.08    0.200000   

       garment_group_no  
0              0.166667  
1              0.041667  
2              0.041667  
3              0.166667  
4              0.041667  
...                 ...  
22028          0.416667  
22029          0.625000  
22030          0.375000  
22031          0.000000  
22032          0.166667  

[22033 rows x 10 columns]

In [349]:
train[['customer_id',
      'article_id',
      'product_code',
      'product_type_no',
      'graphical_appearance_no',
      'colour_group_code',
      'perceived_colour_value_id',
      'perceived_colour_master_id',
      'department_no',
      'index_group_no',
      'section_no',
      'garment_group_no']]

customer_id  article_id  \
0      6793980a0733db0db2b37ee0d63dcdbca67d68e951a359...   693322006   
1      2c4096d75492f8150cd724468d82dfad4deb74c82ebdff...   803757001   
2      a406b87b55ec5be6dceb89de10a668719637d44e2a8971...   611522006   
3      46ef71557837f56d4655cef3c5be315a8496886ec72512...   657510001   
4      54cf679d5f790d76991d110d688d3ec40bd6b049edf80b...   316441005   
...                                                  ...         ...   
22028  330268eadb6b1b1a95876b6aab3e8b6b94fdafdd75c1bf...   663851002   
22029  f89ef8e2b20fd96ec23a89bfca01a04c1aae5b9aa86830...   724906003   
22030  45977583c4d69a0d5c843d96a7d375a75dc37af88c9864...   712216005   
22031  f25575d1bb6daf5835926aa5b06c9dc55e54ee576a8298...   719201002   
22032  41ea93e75936cf7b18251d420037216d38ec868b7ea34d...   730609001   

       product_code  product_type_no  graphical_appearance_no  \
0          0.697507         0.515947                 0.999982   
1          0.829283         0.512195                 0.999988   
2          0.599900         0.512195                 0.999998   
3          0.654774         0.474672                 0.999988   
4          0.247796         0.478424                 0.999988   
...             ...              ...                      ...   
22028      0.662341         0.487805                 0.999976   
22029      0.735194         0.512195                 0.999995   
22030      0.720052         0.476548                 0.999988   
22031      0.728387         0.517824                 0.999995   
22032      0.741999         0.480300                 0.999977   

       colour_group_code  perceived_colour_value_id  \
0               0.085106                      0.375   
1               0.106383                      0.625   
2               0.106383                      0.625   
3               0.106383                      0.625   
4               0.117021                      0.500   
...                  ...                        ...   
22028           0.021277                      0.875   
22029           0.765957                      0.250   
22030           0.117021                      0.500   
22031           0.117021                      0.500   
22032           0.117021                      0.500   

       perceived_colour_master_id  department_no  index_group_no  section_no  \
0                        0.619048       0.808944            1.00    0.031579   
1                        0.285714       0.054051            0.00    0.147368   
2                        0.285714       0.765703            0.12    0.810526   
3                        0.285714       0.049954            0.04    0.536842   
4                        0.476190       0.054051            0.00    0.147368   
...                           ...            ...             ...         ...   
22028                    0.714286       0.734638            0.12    0.463158   
22029                    0.142857       0.064975            0.04    0.578947   
22030                    0.476190       0.036527            0.00    0.136842   
22031                    0.476190       0.208125            0.00    0.136842   
22032                    0.476190       0.528789            0.08    0.200000   

       garment_group_no  
0              0.166667  
1              0.041667  
2              0.041667  
3              0.166667  
4              0.041667  
...                 ...  
22028          0.416667  
22029          0.625000  
22030          0.375000  
22031          0.000000  
22032          0.166667  

[22033 rows x 12 columns]

In [341]:
train[chk_cols]

product_code  product_type_no  graphical_appearance_no  \
0          0.697507         0.515947                 0.999982   
1          0.829283         0.512195                 0.999988   
2          0.599900         0.512195                 0.999998   
3          0.654774         0.474672                 0.999988   
4          0.247796         0.478424                 0.999988   
...             ...              ...                      ...   
22028      0.662341         0.487805                 0.999976   
22029      0.735194         0.512195                 0.999995   
22030      0.720052         0.476548                 0.999988   
22031      0.728387         0.517824                 0.999995   
22032      0.741999         0.480300                 0.999977   

       colour_group_code  perceived_colour_value_id  \
0               0.085106                      0.375   
1               0.106383                      0.625   
2               0.106383                      0.625   
3               0.106383                      0.625   
4               0.117021                      0.500   
...                  ...                        ...   
22028           0.021277                      0.875   
22029           0.765957                      0.250   
22030           0.117021                      0.500   
22031           0.117021                      0.500   
22032           0.117021                      0.500   

       perceived_colour_master_id  department_no  index_group_no  section_no  \
0                        0.619048       0.808944            1.00    0.031579   
1                        0.285714       0.054051            0.00    0.147368   
2                        0.285714       0.765703            0.12    0.810526   
3                        0.285714       0.049954            0.04    0.536842   
4                        0.476190       0.054051            0.00    0.147368   
...                           ...            ...             ...         ...   
22028                    0.714286       0.734638            0.12    0.463158   
22029                    0.142857       0.064975            0.04    0.578947   
22030                    0.476190       0.036527            0.00    0.136842   
22031                    0.476190       0.208125            0.00    0.136842   
22032                    0.476190       0.528789            0.08    0.200000   

       garment_group_no  
0              0.166667  
1              0.041667  
2              0.041667  
3              0.166667  
4              0.041667  
...                 ...  
22028          0.416667  
22029          0.625000  
22030          0.375000  
22031          0.000000  
22032          0.166667  

[22033 rows x 10 columns]

In [314]:
model_knn.fit(df_cust_articles)

NearestNeighbors(algorithm='brute', metric='cosine')

In [315]:
#query_index = np.random.choice(df_cust_articles.shape[0])
query_index = 3

In [316]:
df_cust_articles.shape[0]

22033

In [317]:
print(query_index)

3


In [318]:
distances, indices = model_knn.kneighbors(df_cust_articles.iloc[query_index,:].values.reshape(1, -1), n_neighbors = 5)

In [319]:
df_cust_articles.iloc[query_index,:].values.reshape(1, -1)

array([[0.65477441, 0.47467167, 0.99998812, 0.10638298, 0.625     ,
        0.28571429, 0.04995448, 0.04      , 0.53684211, 0.16666667]])

In [320]:
indices

array([[    3,  6933, 15278, 13152, 20092]], dtype=int64)

In [321]:
df_cust_articles.head()

product_code  product_type_no  graphical_appearance_no  colour_group_code  \
0      0.697507         0.515947                 0.999982           0.085106   
1      0.829283         0.512195                 0.999988           0.106383   
2      0.599900         0.512195                 0.999998           0.106383   
3      0.654774         0.474672                 0.999988           0.106383   
4      0.247796         0.478424                 0.999988           0.117021   

   perceived_colour_value_id  perceived_colour_master_id  department_no  \
0                      0.375                    0.619048       0.808944   
1                      0.625                    0.285714       0.054051   
2                      0.625                    0.285714       0.765703   
3                      0.625                    0.285714       0.049954   
4                      0.500                    0.476190       0.054051   

   index_group_no  section_no  garment_group_no  
0            1.00    0.031579          0.166667  
1            0.00    0.147368          0.041667  
2            0.12    0.810526          0.041667  
3            0.04    0.536842          0.166667  
4            0.00    0.147368          0.041667

In [322]:
# What does this do?
ind = df_cust_articles.index[query_index]

In [323]:
ind

3

In [324]:
df_cust_articles.index[query_index]

3

In [325]:
df_cust_articles.loc[query_index]

product_code                  0.654774
product_type_no               0.474672
graphical_appearance_no       0.999988
colour_group_code             0.106383
perceived_colour_value_id     0.625000
perceived_colour_master_id    0.285714
department_no                 0.049954
index_group_no                0.040000
section_no                    0.536842
garment_group_no              0.166667
Name: 3, dtype: float64

In [326]:
df_cust_articles.loc[query_index,['product_code','product_type_no','graphical_appearance_no']]

product_code               0.654774
product_type_no            0.474672
graphical_appearance_no    0.999988
Name: 3, dtype: float64

In [327]:
train.loc[query_index,['product_code','product_type_no','graphical_appearance_no']]

product_code               0.654774
product_type_no            0.474672
graphical_appearance_no    0.999988
Name: 3, dtype: object

In [328]:
for i in range(0, len(distances.flatten())):
    if i == 0:
        print('Recommendations for {0}:\n'.format(df_cust_articles.index[query_index])) # Again - what does this do?
    else:
        # What does the flatten do?
        print('{0}: Index: {1} {2} with distance of {3}:'.
              format(i,  
                     df_cust_articles.index[indices.flatten()[i]], 
                     df_cust_articles.loc[indices.flatten()[i],['product_code']],
                     distances.flatten()[i]))

Recommendations for 3:

1: Index: 6933 product_code    0.654506
Name: 6933, dtype: float64 with distance of 1.210815236962759e-08:
2: Index: 15278 product_code    0.654506
Name: 15278, dtype: float64 with distance of 1.210815236962759e-08:
3: Index: 13152 product_code    0.648096
Name: 13152, dtype: float64 with distance of 8.811443505618577e-06:
4: Index: 20092 product_code    0.648095
Name: 20092, dtype: float64 with distance of 1.1409569761422489e-05:


In [329]:
for i in range(0, len(distances.flatten())):
    if i == 0:
        print('Recommendations for {0}:\n'.format(train.index[query_index]))
    else:
        print('{0}: Index: {1} {2} {3} {4}, with distance of {5}:'
              .format(i,  
                      train.index[indices.flatten()[i]],
                      train.loc[indices.flatten()[i],['customer_id']], 
                      train.loc[indices.flatten()[i],['article_id']], 
                      train.loc[indices.flatten()[i],['product_code']], 
                      distances.flatten()[i]))

Recommendations for 3:

1: Index: 6933 customer_id    faa03d77b0a88e44d5a3e04cd2246ac91bbfece0cc7f53...
Name: 6933, dtype: object article_id    657285001
Name: 6933, dtype: object product_code    0.654506
Name: 6933, dtype: object, with distance of 1.210815236962759e-08:
2: Index: 15278 customer_id    97a668df609ad74d585c8344ee70a5d5efb76a65c33cb6...
Name: 15278, dtype: object article_id    657285001
Name: 15278, dtype: object product_code    0.654506
Name: 15278, dtype: object, with distance of 1.210815236962759e-08:
3: Index: 13152 customer_id    06d4f1bfd3899633b7d1882b5925f6d76ecbe7f8b4d60c...
Name: 13152, dtype: object article_id    651913003
Name: 13152, dtype: object product_code    0.648096
Name: 13152, dtype: object, with distance of 8.811443505618577e-06:
4: Index: 20092 customer_id    bb70ec803efe3923343d70df1f07ed5f54bdbdb0903c67...
Name: 20092, dtype: object article_id    651912002
Name: 20092, dtype: object product_code    0.648095
Name: 20092, dtype: object, with distanc

###### 

###### 

###### 

## Submission

#### The solution file to have 1371980 prediction rows - use Sample Submission to construct  <a class="tocSkip">

In [333]:
# Read in Sample Submission File
df_sub  = pd.read_csv('../input/sample_submission.csv',
                            #usecols= ['customer_id'], 
                            dtype={'customer_id': 'string'})

In [334]:
numCustomers = df_sub.shape[0]

In [335]:
# Temp code to build submission framework
df_sub_sample = df_sub.sample(frac=0.001, replace=False)
uniBin = 1
listUniBins = [1]

In [336]:
# Set up folder for submission file with predictions
submission_path = '../submission'

In [337]:
# Create folder above project directory
os.makedirs(submission_path, exist_ok=True)  

In [338]:
# Wrtite submission file to disk
df_sub_sample.to_csv(submission_path+'/submission_' + str(uniBin) + '.csv',index=False)

In [339]:
# Check submission file size is correct
for i, uniBin in enumerate(listUniBins):
    dfTemp  = pd.read_csv(submission_path+'/submission_' + str(uniBin) + '.csv')
    if i == 0:
        dfSub = dfTemp
    else:
        dfSub = cudf.concat([dfSub, dfTemp], axis=0)

[print("Matching the number of dfSub rows to the correct submission size. \nSubmission Row: {} v Expected Rows: {}"
                                                                                   .format(dfSub.shape[0],numCustomers))]

Matching the number of dfSub rows to the correct submission size. 
Submission Row: 1372 v Expected Rows: 1371980


[None]

In [340]:
# Check format of submission file
dfCheck = pd.read_csv(submission_path+'/submission_' + str(uniBin) + '.csv')
dfCheck.head()

customer_id  \
0  a818fd7306a747ecef3e874ae4cfa78cb5c8abdc4273ec...   
1  3cb7ae0e626a709828d23c3ef84011bcf160bf5a8a63cd...   
2  613cfe95760cabcb77618e3336b4474b307e6dcbf368f1...   
3  560c07d96be224620c4232fd60440908ef8eeb78ca7938...   
4  9998c2405445468d73a4df096a5d2c237095233d6be360...   

                                          prediction  
0  0706016001 0706016002 0372860001 0610776002 07...  
1  0706016001 0706016002 0372860001 0610776002 07...  
2  0706016001 0706016002 0372860001 0610776002 07...  
3  0706016001 0706016002 0372860001 0610776002 07...  
4  0706016001 0706016002 0372860001 0610776002 07...

###### 

###### 

## References

#### <a class="tocSkip">